include/kernel/sporadic.h

```diff
@@ -34,6 +34,13 @@
 #define MIN_BUDGET_US (2u * getKernelWcetUs() * CONFIG_KERNEL_WCET_SCALE)
 #define MIN_BUDGET    (2u * getKernelWcetTicks() * CONFIG_KERNEL_WCET_SCALE)
 
+/* By ensuring that the total budget of a scheduling context is at least
```

Comment:
only reviewing based on commit message: how can you ensure the first split is always MIN_BUDGET? In the original design, I reasoned that the minimum refills is always 2 and the first split (tail) can be < MIN_BUDGET. Does this result in consuming the budget faster by forcing the initial split to be MIN_BUDGET? 
Note that MIN_BUDGET is based on kernel WCET which doesn't always get used (rarely, in fact) so we shouldn't force that to be minimal consumption.
The original invariant was that the HEAD must be min budget.

Reply:
Just picked up on the TWICE too. TWICE minbudget for a refill is quite large as a minimum consumed value.

Reply:
We somehow ended up with the invariant that all refills should be MIN_BUDGET. I can see why the invariant only needs to be on the head. I also only found the PR where you implement that last week.

This would consume time faster, but it does so by removing time at the end of the refill and delaying it. The effect would be that blocking (without donation) or being interrupted early would delay time at the end of the refill, although that time would still fall within the same period window.

Reply:
can you justify why consuming time faster is appropriate? I'd really like to see the impact in benchmarks before accepting this change.

Reply:
I wasn't saying that it was appropriate or justified, just clarifying what the behaviour there is.

Reply:
I'm not convinced this commit should be merged. If threads always consume MIN_BUDGET I think they will use up their budget unreasonably fast. Also the delta on the algorithm is really large, so I'd need to commit several hours to reviewing this commit properly.